### PR TITLE
Fix grad scaling in Tensor Parallelism

### PR DESCRIPTION
**Pull request by nathan-az**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

[Please link to any issues this PR addresses.](https://github.com/pytorch/torchtune/issues/2641)

#### Changelog
What are the changes made in this PR?
* Forcefully disallows using `foreach` grad scaling implementation with TP
* If not using `foreach`, check grad type, and do not use inplace `mul_` on DTensor
* Rescale grads by world size, not DP size (not sure why this is required but it aligned loss and grad norms across all distributed setups)

image

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/training/_grad_scaler.py [hunk 1]
```diff
@@ -102,4 +103,7 @@ def _scale_grad_(
         else:
             scaler_device = scaler.to(device)
             for g in device_grads:
-                g.mul_(scaler_device)
+                if isinstance(g, DTensor):
```
Comment (joecummings) [inline torchtune/training/_grad_scaler.py:106]: @IvanKobzarev Any idea on why this would be happening?

**Comment by nathan-az** [comment]
@joecummings assuming this (bug + fix) can be repro'd on your side, should we merge this in the short term? Just conscious that TP is recommended for at least one recipe (LLaMA-4) and currently appears incorrect, and a fix in Core for the in-place mul will probably take longer. Can revert in future.

**Review by ebsmothers — APPROVED** [review]
Thanks @nathan-az for figuring this out! I agree we should just land this to unblock TP training. Will merge once CI is green. Also cc @IvanKobzarev since we were talking about this earlier
